extensions/_states/heroku.py

```diff
@@ -0,0 +1,181 @@
+           'changes': {}}
+
+    try:
+        app_config_vars = __salt__['heroku.list_app_config_vars'](name, api_key)
+        log.debug("Calling _diff_heroku_pillar_vars")
+        _diff_heroku_pillar_vars(app_config_vars, config_vars)
+        if all(not v for v in diff_heroku_pillar_dict.values()):
+            ret['result'] = True
+            ret['comment'] = 'No changes detected'
+        else:
+            ret['result'] = False
+            ret['comment'] = 'Pillar and Heroku mismatch'
+            ret['changes'] = {'old': {'heroku': heroku_diff_dict, 'pillar': pillar_diff_dict}}
+    except:
+        e = sys.exc_info()[0]
+        ret['result'] = False
+        log.exception(e)
+
+    return ret
+
+
+def _diff_heroku_pillar_vars(app_config_vars, config_vars):
+    '''
+    Internal function that runs multiple set operations on Heroku and Pillar
+    dicts and sets the values on diff_heroku_pillar_dict accordingly
+
+    :param app_config_vars: The Heroku configuration variables
+    :param config_vars: The Pillar configuration variables
+    :returns: three global dicts
+    '''
+
+
+    global diff_heroku_pillar_dict, heroku_diff_dict, pillar_diff_dict
+    app_config_vars_set = set(app_config_vars.items())
+    config_vars_set = set(config_vars.items())
+
+    if app_config_vars_set.symmetric_difference(config_vars_set):
+        diff_heroku_pillar_dict['key_or_value_mismatch'] = True
+        heroku_diff_dict = dict(app_config_vars_set.difference(config_vars_set))
+        pillar_diff_dict = dict(config_vars_set.difference(app_config_vars_set))
+    if len(app_config_vars) - len(config_vars) > 0:
+        diff_heroku_pillar_dict['heroku_not_pillar'] = True
+    if len(app_config_vars) - len(config_vars) < 0:
+        diff_heroku_pillar_dict['pillar_not_heroku'] = True
+
+    log.debug("Completed _diff_heroku_pillar_vars")
+    return diff_heroku_pillar_dict, heroku_diff_dict, pillar_diff_dict
+
+
+def update_app_config_vars(name, config_vars, api_key):
+    '''
+    Update Heroku config vars based on specified pillar config vars.
+    This function will perform the following actions:
+      - Change Heroku config var values for the keys that match one in pillar
+      - Add keys from pillar to Heroku
+      - Leave Heroku keys not find in pillar unchanged
+
+    :param name: The name of the Heroku app
+    :param config_vars: The config_vars specified in pillar
+    :param api_key: The Heroku api_key
+    :returns: Result of executing the state
+    :rtype: dict
+    '''
+
+
+    ret = {'name': name,
+           'comment': '',
+           'result': True,
+           'changes': {}}
+
+    diff_app_config_vars(name, config_vars, api_key)
+    if any(v for v in diff_heroku_pillar_dict.values()):
+        __salt__['heroku.update_app_config_vars'](name, config_vars, api_key)
+        new_app_config_vars = __salt__['heroku.list_app_config_vars'](name, api_key)
+        ret['changes']['new'] = {'heroku': new_app_config_vars}
```

Comment:
It doesn't appear that you're setting `old` anywhere for the ret object

Reply:
It's being used [here](https://github.com/mitodl/salt-extensions/blob/2ceb1cb437ce5fdb555417f44b1d73b4a43edb4c/extensions/_states/heroku.py#L73) and [here](https://github.com/mitodl/salt-extensions/blob/2ceb1cb437ce5fdb555417f44b1d73b4a43edb4c/extensions/_states/heroku.py#L168)